### PR TITLE
S3 Key Resource

### PR DESCRIPTION
**Pull request by Krishnan-Priyanshu**
# Description
A few sentences describing the overall goals of the pull request's commits.
image

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| |

# ISSUE TYPE

- Docs Pull Request
- Feature Pull Request
- Test Pull Request

##### OUTPUT

```paste below

```
##### ADDITIONAL INFORMATION

```paste below

```
# Checklist:

- [ ] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 80% code coverage
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have maintained backward compatibility

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [ ] Unit tests
- [ ] Acceptance tests

### CODE REVIEW

**Comment by terraform-providers-svc** [comment]
Can one of the admins verify this patch?

**Comment by Krishnan-Priyanshu** [comment]
run e2e test TestAccS3Key

File: powerscale/provider/s3_key_resource.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+func S3KeyResourceSchema() map[string]schema.Attribute {
+	return map[string]schema.Attribute{
+		"access_id": schema.StringAttribute{
+			Computed:            true,
+			MarkdownDescription: "Unique identifier of the S3 key.",
+			Description:         "Unique identifier of the S3 key.",
+		},
+		"user": schema.StringAttribute{
+			Required:            true,
+			MarkdownDescription: "The username to create the S3 key. Required.",
```
Comment (shenda1) [inline powerscale/provider/s3_key_resource.go:93]: Do we need to specify Required explicitly?

File: powerscale/provider/s3_key_resource.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+	}
+	if resp.Keys.GetSecretKeyTimestamp() != int32(s3key.SecretKeyTimestamp.ValueInt64()) {
+		response.Diagnostics.AddWarning(errMsg, errMsg)
+		s3key.SecretKey = types.StringValue(errMsg)
+	}
+	err = helper.CopyFieldsToNonNestedModel(ctx, resp.Keys, &s3key)
+	if err != nil {
+		response.Diagnostics.AddError("Error reading s3 key ", err.Error())
+	}
+	response.State.Set(ctx, s3key)
```
Comment (shenda1) [inline powerscale/provider/s3_key_resource.go:200]: What if it fails to set the state?

File: powerscale/provider/s3_key_resource.go [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+	// call get s3key
+	resp, err := helper.GetS3Key(ctx, r.client, s3key)
+	if err != nil {
+		response.Diagnostics.AddError("Error reading s3 key ", err.Error())
+		return
+	}
+
+	// precheck to invalidate the refresh
+	errMsg := "[UNKNOWN KEY] Key Generated Outside of Terraform"
+	if resp.Keys.GetOldKeyTimestamp() != int32(s3key.SecretKeyTimestamp.ValueInt64()) {
```
Comment (shenda1) [inline powerscale/provider/s3_key_resource.go:187]: Can we add the comments ?

File: powerscale/provider/s3_key_resource.go (hunk 2 above)
Comment (Krishnan-Priyanshu) [reply powerscale/provider/s3_key_resource.go:200]: added the return inside if error block

File: powerscale/provider/s3_key_resource.go (hunk 3 above)
Comment (Krishnan-Priyanshu) [reply powerscale/provider/s3_key_resource.go:187]: added the comment

File: docs/resources/s3_key.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
+---
+# Copyright (c) 2023-2024 Dell Inc., or its subsidiaries. All Rights Reserved.
```
Comment (Krunal-Thakkar) [inline docs/resources/s3_key.md:2]: only 2024

File: powerscale/provider/s3_key_resource.go (hunk 2 above)
Comment (shenda1) [reply powerscale/provider/s3_key_resource.go:203]: response.Diagnostics.Append(response.State.Set(ctx, s3key)) ?

**Review by shenda1 — COMMENTED** [review]
Add the coverage report

**Comment by Krishnan-Priyanshu** [comment]
> Add the coverage report

Coverage Report around 88%, attached inside PR Description

File: powerscale/provider/s3_key_resource.go (hunk 2 above)
Comment (Krishnan-Priyanshu) [reply powerscale/provider/s3_key_resource.go]: added

**Review by shenda1 — APPROVED** [review]

**Review by Krunal-Thakkar — APPROVED** [review]

**Comment by Krunal-Thakkar** [comment]
run e2e test TestAccS3Key

**Comment by Krishnan-Priyanshu** [comment]
run e2e test S3Key

**Comment by Krishnan-Priyanshu** [comment]
run e2e test S3Key

**Comment by Krishnan-Priyanshu** [comment]
run e2e test S3Key

**Comment by Krishnan-Priyanshu** [comment]
run e2e test S3Key

**Comment by Krishnan-Priyanshu** [comment]
run e2e test S3Key

**Comment by Krishnan-Priyanshu** [comment]
run e2e test S3Key

**Comment by Krishnan-Priyanshu** [comment]
run e2e test S3Key

**Comment by Krishnan-Priyanshu** [comment]
run e2e test S3Key

**Pull request merged**